PLAT-1780 Remove dependency on django-extensions

jmbowman opened the pull request:
Removes our dependency on django-extensions, which was breaking seemingly unrelated migrations whenever attempting to upgrade to a version new enough to work with Django 1.11.  We were only using the following functionality from it:

* `TimeStampedModel`, which has an exact equivalent in `django-model-utils` that we use far more often.  I switched the handful of uses of this over to that implementation instead (no new migrations required).
* One direct usage of `CreationDateTimeField`, which again has an equivalent in `django-model-utils`.  Switched it over, and again no new migration is needed.
* The `print_settings` management command, used only to get the value of `STATIC_ROOT` from the Django settings for some paver commands.  Replaced this with a new, simpler `print_setting` management command.
* The `reset_db` management command, used to drop and recreate a database during some test runs.  I copied in a slightly simplified version of this management command.

jmbowman commented:
jenkins run python

jmbowman commented:
A prior run of the unit tests had a single test failure, which I didn't see occurring in the Splunk logs for the past month: https://build.testeng.edx.org/job/edx-platform-test-subset/967839/consoleFull .  Additional eyeballs on that would be appreciated, I don't like to ignore flakes like that; seems to have gone down some path where the catalog API endpoint didn't get mocked correctly.  Haven't been able to reproduce the failure locally yet.

bmedx commented on openedx/core/djangoapps/extensions_lite/management/commands/reset_db.py:
```diff
@@ -0,0 +1,190 @@
+            return user, password, database_name, database_host, database_port
+
+        except configparser.NoSectionError:
+            pass
```
I would expect that if the conf is included in settings it should be usable here. Maybe log it if things fall through here?

bmedx commented on openedx/core/djangoapps/site_configuration/models.py:
```diff
@@ -130,6 +130,10 @@ class SiteConfigurationHistory(TimeStampedModel):
 
+    class Meta:
+        get_latest_by = 'modified'
+        ordering = ('-modified', '-created',)
```
Probably not an issue here, but worth noting that those cols don't seem to be indexed. Only 130 odd rows in the read replica (and 285 for ApiAccessRequest), but thought I'd bring it up. I also happened to see an "api_admin_historicalapiraccessequest" table in the read replica that I think is an orphan?

jmbowman replied:
These got silently added to the model definition by the previous implementation of TimeStampedModel, and I didn't want to risk breaking something by removing it at this time.  Also, removing it now would require a new migration (although one that shouldn't actually perform any database changes).  It might merit a separate ticket to review all these and see if it's worth adding indexes or removing the default sorting, though.

The orphaned table is one of the ones that should (eventually) get cleaned up by [PLAT-1756](https://openedx.atlassian.net/browse/PLAT-1756).

jmbowman commented on openedx/core/djangoapps/extensions_lite/__init__.py:
```diff
@@ -0,0 +1,7 @@
+Replicated here because django-extensions was proving difficult to keep up to
+date thanks to its large surface area for bugs and Django version
+incompatibility.
+"""
```
That might work; I hadn't spotted it because there are no management commands currently in it.  I think @macdiesel recently had some import conflict problems with that one, but it shouldn't be a problem for loading management commands.

pwnage101 commented on lms/djangoapps/certificates/migrations/0001_initial.py:
```diff
@@ -116,7 +115,7 @@ class Migration(migrations.Migration):
                 ('course_id', CourseKeyField(default=None, max_length=255, blank=True)),
                 ('whitelist', models.BooleanField(default=0)),
-                ('created', django_extensions.db.fields.CreationDateTimeField(auto_now_add=True, verbose_name='created')),
+                ('created', model_utils.fields.AutoCreatedField(default=django.utils.timezone.now, verbose_name='created', editable=False)),
```
Looks like the default is already `django.utils.timezone.now`, does this need to be explicit if "Created" is already in the class name?

pwnage101 replied:
oh, is this an automatically generated file?

macdiesel replied:
Correct, this file is generated by makemigrations command.

jmbowman replied:
To be specific, the file was automatically generated back when we were using django-extensions (which generated the line being replaced here).  I'm replacing that line with the output that has been generated in other migration files where we were already using `AutoCreatedField`.  Rewriting history like this is ok in this case because both lines yield the exact same database schema.  There will be some subtle differences in exactly when the timestamp gets created and what happens if you try to change the timestamp after the model instance has already been created and saved, but it's extremely unlikely that we have any code that really cares about those details.

I could have made the behavior match exactly what it was doing before, but then it would be inconsistent with the behavior of all our other `created` and `modified` fields, which use this default behavior from `django-model-utils`.

jmbowman commented:
I moved the management commands to `openedx.core.djangoapps.util` and added some tests for `print_setting` adapted from the `django-extensions` tests.  Any more suggestions for improvement?
